File: emulinker/src/main/java/org/emulinker/util/GameDataQueue.java
package org.emulinker.util;

import java.util.concurrent.*;

public class GameDataQueue {
  private int gameID;
  private int numPlayers;
  private PlayerDataQueue[] playerQueues;
  private int timeoutMillis;
  private int retries;
  private boolean gameDesynched = false;

  public GameDataQueue(int gameID, int numPlayers, int timeoutMillis, int retries) {
    this.gameID = gameID;
    this.numPlayers = numPlayers;
    this.timeoutMillis = timeoutMillis;
    this.retries = retries;

    playerQueues = new PlayerDataQueue[numPlayers];
    for (int i = 0; i < playerQueues.length; i++) playerQueues[i] = new PlayerDataQueue((i + 1));
  }

  public int getGameID() {
    return gameID;
  }

  public int getNumPlayers() {
    return numPlayers;
  }

  public int getTimeoutMillis() {
    return timeoutMillis;
  }

  public int getRetries() {
    return retries;
  }

  public void setGameDesynched() {
    gameDesynched = true;
  }

  public void addData(int playerNumber, byte data[]) {
    for (int i = 0; i < data.length; i++) addData(playerNumber, data[i]);
  }

  public void addData(int playerNumber, byte data) {
    for (int i = 0; i < numPlayers; i++) playerQueues[i].addData(playerNumber, data);
  }

  public byte[] getData(int playerNumber, int byteCount, int bytesPerAction)
      throws PlayerTimeoutException, DesynchException {
    return playerQueues[(playerNumber - 1)].getData(byteCount, bytesPerAction);
  }

  private class PlayerDataQueue {
    private CircularBlockingByteQueue[] queues;
    private int lastI = 0;
    private int lastJ = 0;
    private byte[] lastData = null;
    private int timeoutCounter = 0;

    private PlayerDataQueue(int playerNumber) {
      queues = new CircularBlockingByteQueue[numPlayers];
      for (int i = 0; i < queues.length; i++)
        queues[i] = new CircularBlockingByteQueue(((numPlayers * 6) * 4));
    }

    private void addData(int playerNumber, byte data) {
      queues[(playerNumber - 1)].put(data);
    }

    private byte[] getData(int byteCount, int bytesPerAction)
        throws PlayerTimeoutException, DesynchException {
      byte[] data = null;

      if (lastData != null) {
        data = lastData;
        lastData = null;
        //				log.debug("Player " + thisPlayerNumber + ": getData with i=" + lastI + ", j=" +
        // lastJ);
      } else data = new byte[(byteCount * numPlayers)];

      for (int i = lastI; i < ((byteCount / bytesPerAction) * numPlayers); i++) {
        for (int j = lastJ; j < bytesPerAction; j++) {
          try {
            data[((i * bytesPerAction) + j)] =
                queues[(i % numPlayers)].get(timeoutMillis, TimeUnit.MILLISECONDS);
          } catch (TimeoutException e) {
            lastI = i;
            lastJ = j;
            lastData = data;

            if (++timeoutCounter > retries)
              throw new DesynchException(
                  "Player " + ((i % numPlayers) + 1) + " is lagged!", ((i % numPlayers) + 1), e);
            else throw new PlayerTimeoutException(((i % numPlayers) + 1), timeoutCounter, e);
          }
        }
      }

      lastI = lastJ = 0;
      lastData = null;
      timeoutCounter = 0;

      return data;
    }
  }

  public static class PlayerTimeoutException extends Exception {
    private int playerNumber;
    private int timeoutNumber;

    private PlayerTimeoutException(int playerNumber, int timeoutNumber, TimeoutException e) {
      super(e);
      this.playerNumber = playerNumber;
      this.timeoutNumber = timeoutNumber;
    }

    public int getPlayerNumber() {
      return playerNumber;
    }

    public int getTimeoutNumber() {
      return timeoutNumber;
    }
  }

  public static class DesynchException extends Exception {
    private int playerNumber;

    private DesynchException(String msg, int playerNumber, TimeoutException e) {
      super(msg, e);
      this.playerNumber = playerNumber;
    }

    public int getPlayerNumber() {
      return playerNumber;
    }
  }
}

File: emulinker/src/main/java/org/emulinker/util/GameDataQueue.kt
package org.emulinker.util

import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

class GameDataQueue(
    val gameID: Int, val numPlayers: Int, val timeoutMillis: Int, val retries: Int
) {
  private val playerQueues: Array<PlayerDataQueue?> = arrayOfNulls(numPlayers)
  private var gameDesynched = false

  fun setGameDesynched() {
    gameDesynched = true
  }

  fun addData(playerNumber: Int, data: ByteArray) {
    for (i in data.indices) addData(playerNumber, data[i])
  }

  fun addData(playerNumber: Int, data: Byte) {
    for (i in 0 until numPlayers) playerQueues[i]!!.addData(playerNumber, data)
  }

  @Throws(PlayerTimeoutException::class, DesynchException::class)
  fun getData(playerNumber: Int, byteCount: Int, bytesPerAction: Int): ByteArray? {
    return playerQueues[playerNumber - 1]!!.getData(byteCount, bytesPerAction)
  }

  private inner class PlayerDataQueue internal constructor(playerNumber: Int) {
    private val queues: Array<CircularBlockingByteQueue?> = arrayOfNulls(numPlayers)
    private var lastI = 0
    private var lastJ = 0
    private var lastData: ByteArray? = null
    private var timeoutCounter = 0

    internal fun addData(playerNumber: Int, data: Byte) {
      queues[playerNumber - 1]!!.put(data)
    }

    @Throws(PlayerTimeoutException::class, DesynchException::class)
    fun getData(byteCount: Int, bytesPerAction: Int): ByteArray? {
      val data: ByteArray?
      if (lastData != null) {
        data = lastData
        lastData = null
        //				logger.atFine().log("Player " + thisPlayerNumber + ": getData with i=" + lastI + ",
        // j=" +
        // lastJ);
      } else data = ByteArray(byteCount * numPlayers)
      for (i in lastI until byteCount / bytesPerAction * numPlayers) {
        for (j in lastJ until bytesPerAction) {
          try {
            data!![i * bytesPerAction + j] =
                queues[i % numPlayers]!![timeoutMillis.toLong(), TimeUnit.MILLISECONDS]
          } catch (e: TimeoutException) {
            lastI = i
            lastJ = j
            lastData = data
            if (++timeoutCounter > retries)
                throw DesynchException(
                    "Player " + (i % numPlayers + 1) + " is lagged!", i % numPlayers + 1, e)
            else throw PlayerTimeoutException(i % numPlayers + 1, timeoutCounter, e)
          }
        }
      }
      lastJ = 0
      lastI = lastJ
      lastData = null
      timeoutCounter = 0
      return data
    }

    init {
      for (i in queues.indices) queues[i] = CircularBlockingByteQueue(numPlayers * 6 * 4)
    }
  }

  class PlayerTimeoutException(val playerNumber: Int, val timeoutNumber: Int, e: TimeoutException) :
      Exception(e)
  class DesynchException(msg: String, val playerNumber: Int, e: TimeoutException) :
      Exception(msg, e)

  init {
    for (i in playerQueues.indices) playerQueues[i] = PlayerDataQueue(i + 1)
  }
}
